url4aac7e6f55_simon0191_tetris3d_tgz-pJ8-Tetris3dJ8/src/vsdk/toolkit/io/PersistenceElement.java

```diff
@@ -0,0 +1,734 @@
+
+        return a;
+    }
+
+    /**
+    Given a previously initialized array of bytes, this method fills it
+    with information readed from the given input stream.  If it is not
+    enough information to read, this method generates an Exception.
+    */
+    public static void
+    readBytes(InputStream is, byte[] bytesBuffer) throws Exception
+    {
+        int offset = 0;
+        int numRead = 0;
+        int length = bytesBuffer.length;
+        do {
+            numRead = is.read(bytesBuffer, 
+                              offset, (length-offset));
+            offset += numRead;
+        } while( offset < length && numRead >= 0 ); 
+    }
+
+    /**
+    Given a previously initialized array of bytes, this method writes it
+    with information readed from the given output stream.  If it is not
+    enough information to read, this method generates an Exception.
+    */
+    public static void
+    writeBytes(OutputStream os, byte[] bytesBuffer) throws Exception
+    {
+        os.write(bytesBuffer, 0, bytesBuffer.length);
+    }
+
+    private static void int2byteArrayDirect(byte[] arr, int start, int num)
+    {
+        int i = 0;
+        int len = 2;
+        byte[] tmp = new byte[len];
+
+        // Convert number to array
+        for ( i = 0; i < len; i++ ) {
+            tmp[i] = (byte)((num & (0xFF << 8*i)) >> 8*i);
+        }
+
+        // Export subarray to end array
+        int cnt;
+        for ( i = start, cnt = 0; i < (start + len); i++, cnt++ ) {
+            arr[i] = tmp[cnt];
+        }
+    }
+
+    private static void int2byteArrayInvert(byte[] arr, int start, int num)
+    {
+        int i = 0;
+        int len = 2;
+        byte[] tmp = new byte[len];
+
+        // Convert number to array
+        for ( i = 0; i < len; i++ ) {
+            tmp[len-i-1] = (byte)((num & (0xFF << 8*i)) >> 8*i);
+        }
+
+        // Export subarray to end array
+        int cnt;
+        for ( i = start, cnt = 0; i < (start + len); i++, cnt++ ) {
+            arr[i] = tmp[cnt];
+        }
+    }
+
+    private static int byteArray2intDirect(byte[] arr, int start) {
+        int low = arr[start] & 0xff;
+        int high = arr[start+1] & 0xff;
+        return ( high << 8 | low );
+    }
+
+    private static int byteArray2intInvert(byte[] arr, int start) {
+        int low = arr[start] & 0xff;
+        int high = arr[start+1] & 0xff;
+        return ( low << 8 | high );
+    }
+
+    private static long byteArray2longDirect(byte[] arr, int start) {
+        int i = 0;
+        int len = 4;
+        int cnt = 0;
+        byte[] tmp = new byte[len];
+        for ( i = start; i < (start + len); i++ ) {
+            tmp[cnt] = arr[i];
+            cnt++;
+        }
+        long accum = 0;
+        i = 0;
+        for ( int shiftBy = 0; shiftBy < 32; shiftBy += 8 ) {
+            accum |= ( (long)( tmp[i] & 0xff ) ) << shiftBy;
+            i++;
+        }
+        return accum;
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 180:194, 234:248.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.